python/lsst/ts/attcpip/at_simulator.py

```diff
@@ -0,0 +1,246 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ["AtSimulator"]
+
+import abc
+import logging
+import types
+import typing
+
+import jsonschema
+import numpy as np
+from lsst.ts import tcpip
+
+from .at_server_simulator import AtServerSimulator
+from .enums import Ack, CommonCommandArgument
+from .schemas import registry
+
+
+class AtSimulator:
+    """Common code for simulating an AT system."""
+
+    def __init__(self) -> None:
+        self.log = logging.getLogger(type(self).__name__)
+        # TODO DM-38912 Send "port" via configuration.
+        self.cmd_evt_server = AtServerSimulator(
+            host=tcpip.LOCALHOST_IPV4,
+            port=5000,
+            log=self.log,
+            dispatch_callback=self.cmd_evt_dispatch_callback,
+            connect_callback=self.cmd_evt_connect_callback,
+            name="CmdEvtServer",
+        )
+        # TODO DM-38912 Send "port" via configuration.
+        self.telemetry_server = AtServerSimulator(
+            host=tcpip.LOCALHOST_IPV4,
+            port=6000,
+            log=self.log,
+            dispatch_callback=self.telemetry_dispatch_callback,
+            connect_callback=self.telemetry_connect_callback,
+            name="TelemetryServer",
+        )
+
+        # Keep track of the sequence_id as commands come dripping in. The
+        # sequence ID should raise monotonally without gaps. If a gap is seen,
+        # a NOACK should be returned.
+        self.last_sequence_id: int = 0
+
+        # Dict of command: function.
+        self.dispatch_dict: dict[str, typing.Callable] = {}
+
+        self.load_schemas()
+
+    @abc.abstractmethod
+    def load_schemas(self) -> None:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    async def cmd_evt_connect_callback(self, server: tcpip.OneClientServer) -> None:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    async def cmd_evt_dispatch_callback(self, data: typing.Any) -> None:
+        raise NotImplementedError
+
+    async def telemetry_connect_callback(self, server: tcpip.OneClientServer) -> None:
+        """Callback function for when a telemetry client connects or
+        disconnects."""
+        pass
+
+    async def telemetry_dispatch_callback(self, data: typing.Any) -> None:
+        """Asynchronous function to call when data are read and dispatched.
+
+        The received data ignored since the telemetry server is only supposed
+        to send data and not to receive any.
```

Comment:
I do not understand this sentence for this part: "The received data ignored". Do you try to say "The received data is ignored"?

Reply:
Thanks for catching that. Yes, `is` is missing. I have added it now.